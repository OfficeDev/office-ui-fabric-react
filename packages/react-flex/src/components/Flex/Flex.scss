:root {
  --flex-display: flex;
  --flex-wrap: nowrap;
  --flex-direction: row;
  --flex-verticalAlign: stretch;
  --flex-horizontalAlign: flex-start;
  --flex-padding: 0;
  --flex-gap: 0 0;
}

.root {
  background: var(--flex-background);
  display: var(--flex-display);
  flex-wrap: var(--flex-wrap);
  align-items: var(--flex-verticalAlign);
  flex-direction: var(--flex-direction); // problem
  justify-content: var(--flex-horizontalAlign);
  max-height: var(--flex-maxHeight);
  max-width: var(--flex-maxWidth);
  padding: var(--flex-padding);
  margin: var(--flex-gap);

  &._column {
    --flex-direction: column;
  }

  &._inline {
    --flex-display: inline-flex;
  }

  &._fluid {
    height: 100%;
    max-height: 100%;
  }

  &._wrap {
    --flex-wrap: wrap;
  }

  &._horizontalAlign_start {
    --flex-horizontalAlign: flex-start;
  }

  &._horizontalAlign_end {
    --flex-horizontalAlign: flex-end;
  }

  &._horizontalAlign_center {
    --flex-horizontalAlign: center;
  }

  &._horizontalAlign_between {
    --flex-horizontalAlign: space-between;
  }

  &._horizontalAlign_around {
    --flex-horizontalAlign: space-around;
  }

  &._horizontalAlign_evenly {
    --flex-horizontalAlign: space-evenly;
  }

  &._verticalAlign_start {
    --flex-verticalAlign: flex-start;
  }

  &._verticalAlign_end {
    --flex-verticalAlign: flex-end;
  }

  &._verticalAlign_center {
    --flex-verticalAlign: center;
  }

  &._verticalAlign_stretch {
    --flex-verticalAlign: stretch;
  }

  &._verticalAlign_baseline {
    --flex-verticalAlign: baseline;
  }
}
