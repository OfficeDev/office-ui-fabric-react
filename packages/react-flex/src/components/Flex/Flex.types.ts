import { ComponentProps, BaseSlots, SlotProps } from '@fluentui/react-compose';

/**
 * Defines a type made by the union of the different values that the align-items and justify-content flexbox
 * properties can take.
 */
export type Alignment = 'start' | 'end' | 'center' | 'between' | 'around' | 'evenly' | 'baseline' | 'stretch';

export interface FlexProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /**
   * Defines if container should be inline element.
   */
  inline?: boolean;
  /**
   * Defines whether to render the children horizontally.
   */
  column?: boolean;
  /**
   * Sets vertical flow direction.
   */
  wrap?: boolean;
  /**
   * Defines how to align the children horizontally (along the x-axis).
   */
  horizontalAlign?: Alignment;
  /**
   * Defines how to align the children vertically (along the y-axis).
   */
  verticalAlign?: Alignment;
  /**
   * Defines the inner padding of the container.
   */
  padding?: number | string;
  /**
   * Defines whether to render the children in the opposite direction.
   */
  reversed?: boolean;
  /**
   * Defines the spacing between the children.
   */
  gap?: number | string;
  /**
   * Defines whether the children should not shrink to fit the available space.
   */
  disableShrink?: boolean;
  /**
   * Defines whether the container should take up 100% of the height of its parent.
   */
  fluid?: boolean;
  /**
   * Defines strategy for distributing remaining space between items.
   */
  space?: ['around', 'between', 'evenly'];
}

export type RecursivePartial<T> = {
  [P in keyof T]?: T[P] extends (infer U)[]
    ? RecursivePartial<U>[]
    : T[P] extends object
    ? RecursivePartial<T[P]>
    : T[P];
};

export interface FlexSlots extends BaseSlots {}

export type FlexSlotProps = SlotProps<FlexSlots, FlexProps, React.HTMLAttributes<HTMLElement>>;
