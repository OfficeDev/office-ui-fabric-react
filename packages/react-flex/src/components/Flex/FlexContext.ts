import * as React from 'react';

type FlexContextProps = {
  disableShrink: boolean;
  gap: undefined | string;
};

const initialState: FlexContextProps = {
  disableShrink: false,
  gap: undefined,
};

export const FlexContext = React.createContext(initialState);
