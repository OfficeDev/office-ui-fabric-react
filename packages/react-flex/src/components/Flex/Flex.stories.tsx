import * as React from 'react';
import { Flex } from './Flex';
import { FlexItem } from '../FlexItem';
import * as classes from './Flex.stories.scss';

export const inlineFlex = () => (
  <Flex inline={true}>
    <FlexItem>
      <p>
        Hello eeeee
        <Flex inline={true}>
          <FlexItem>
            <h2 className={classes.text}>Hello World</h2>
          </FlexItem>
          <FlexItem>
            <h2 className={classes.text}>Goodbye World</h2>
          </FlexItem>
          <FlexItem>
            <h2 className={classes.text}>I'm back World</h2>
          </FlexItem>
        </Flex>
        eeeeeeeeeee
      </p>
    </FlexItem>
    <FlexItem>
      <p>
        Hello eeeee
        <Flex inline={false}>
          <FlexItem>
            <h2 className={classes.text}>Hello World</h2>
          </FlexItem>
          <FlexItem>
            <h2 className={classes.text}>Goodbye World</h2>
          </FlexItem>
          <FlexItem>
            <h2 className={classes.text}>I'm back World</h2>
          </FlexItem>
        </Flex>
        eeeeeeeeeee
      </p>
    </FlexItem>
  </Flex>
);
