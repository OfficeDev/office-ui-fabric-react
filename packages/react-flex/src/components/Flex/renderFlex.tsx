/* eslint-disable react/no-danger */
import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { FlexState } from './Flex.types';

/**
 * Render the final JSX of Flex
 * {@docCategory Flex }
 */
export const renderFlex = (state: FlexState) => {
  const { slots, slotProps } = getSlots(state);

  const direction = state.direction ? state.direction : 'row';
  const horizontalAlign = state.horizontalAlign ? state.horizontalAlign : 'normal';
  const verticalAlign = state.verticalAlign ? state.verticalAlign : 'normal';

  return (
    <>
      <style
        dangerouslySetInnerHTML={{
          __html: `
          .test {
            display: ${state.inline ? 'inline-flex' : 'flex'};
            flex-direction: ${direction};
            flex-wrap: ${state.wrap ? 'wrap' : 'nowrap'};
            justify-content: ${direction.startsWith('row') ? horizontalAlign : verticalAlign};
            align-items: ${direction.startsWith('row') ? verticalAlign : horizontalAlign};
          }

          .test > * {
            margin: ${state.gap === undefined ? 0 : state.gap};
            order: 0;
            flex-grow: ${state.grow === undefined ? 0 : state.grow};
            flex-shrink: ${state.shrink === undefined ? 1 : state.shrink};
            flex-basis: auto;
            align-self: auto;
          }`,
        }}
      />
      <slots.root {...slotProps.root} className="test">
        {state.children}
      </slots.root>
    </>
  );
};
