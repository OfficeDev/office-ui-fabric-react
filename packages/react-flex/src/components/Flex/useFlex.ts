import * as React from 'react';
import { makeMergeProps } from '@fluentui/react-utilities';
import { FlexProps, FlexState } from './Flex.types';

const mergeProps = makeMergeProps<FlexState>();

/**
 * Create the state required to render Flex.
 *
 * The returned state can be modified with hooks such as useFlexStyles,
 * before being passed to renderFlex.
 *
 * @param props - props from this instance of Flex
 * @param ref - reference to root HTMLElement of Flex
 * @param defaultProps - (optional) default prop values provided by the implementing type
 *
 * {@docCategory Flex }
 */
export const useFlex = (props: FlexProps, ref: React.Ref<HTMLElement>, defaultProps?: FlexProps): FlexState => {
  const state = mergeProps(
    {
      ref,
    },
    defaultProps,
    props,
  );

  return state;
};
