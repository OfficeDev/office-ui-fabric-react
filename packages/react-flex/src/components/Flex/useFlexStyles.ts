import { mergeClasses, makeStyles } from '@fluentui/react-make-styles';
import { FlexState } from './Flex.types';

const cssVars = {
  gap: '--gap',
  grow: '--grow',
  shrink: '--shrink',
};

const useStyles = makeStyles({
  root: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'nowrap',
    justifyContent: 'flex-start',
    alignItems: 'stretch',
    [cssVars.gap]: 0,
    [cssVars.grow]: 0,
    [cssVars.shrink]: 1,
    '> *': {
      order: 0,
      margin: `var(${cssVars.gap})`,
      flexGrow: `var(${cssVars.grow})`,
      flexShrink: `var(${cssVars.shrink})`,
      flexBasis: 'auto',
      alignSelf: 'auto',
    },
  },

  flexDirectionRow: { flexDirection: 'row' },
  flexDirectionRowReverse: { flexDirection: 'row-reverse' },
  flexDirectionColumn: { flexDirection: 'column' },
  flexDirectionColumnReverse: { flexDirection: 'column-reverse' },

  justifyContentFlexStart: { justifyContent: 'flex-start' },
  justifyContentFlexEnd: { justifyContent: 'flex-end' },
  justifyContentCenter: { justifyContent: 'center' },
  justifyContentSpaceBetween: { justifyContent: 'space-between' },
  justifyContentSpaceAround: { justifyContent: 'space-around' },

  alignItemsFlexStart: { alignItems: 'flex-start' },
  alignItemsFlexEnd: { alignItems: 'flex-end' },
  alignItemsCenter: { alignItems: 'center' },
  alignItemsBaseline: { alignItems: 'baseline' },
  alignItemsStretch: { alignItems: 'stretch' },

  flexWrap: { flexWrap: 'wrap' },
  flexNoWrap: { flexWrap: 'nowrap' },

  displayInlineFlex: { display: 'inline-flex' },
  displayFlex: { display: 'flex' },
});

/**
 * Apply styling to the Flex slots based on the state
 * {@docCategory Flex }
 */
export const useFlexStyles = (state: FlexState): FlexState => {
  const styles = useStyles();

  const isRowDirection = state.direction?.substring(0, 3) === 'row';
  state.className = mergeClasses(
    styles.root,
    state.direction === 'row' && styles.flexDirectionRow,
    state.direction === 'column' && styles.flexDirectionColumn,
    state.direction === 'row-reverse' && styles.flexDirectionRowReverse,
    state.direction === 'column-reverse' && styles.flexDirectionColumnReverse,
    state.horizontalAlign === 'flex-start' &&
      (isRowDirection ? styles.justifyContentFlexStart : styles.alignItemsFlexStart),
    state.verticalAlign === 'flex-start' &&
      (isRowDirection ? styles.alignItemsFlexStart : styles.justifyContentFlexStart),
    state.horizontalAlign === 'flex-end' && (isRowDirection ? styles.justifyContentFlexEnd : styles.alignItemsFlexEnd),
    state.verticalAlign === 'flex-end' && (isRowDirection ? styles.alignItemsFlexEnd : styles.justifyContentFlexEnd),
    state.horizontalAlign === 'center' && (isRowDirection ? styles.justifyContentCenter : styles.alignItemsCenter),
    state.verticalAlign === 'center' && (isRowDirection ? styles.alignItemsCenter : styles.justifyContentCenter),
    state.horizontalAlign === 'space-around' && isRowDirection && styles.justifyContentSpaceAround,
    state.verticalAlign === 'space-around' && !isRowDirection && styles.justifyContentSpaceAround,
    state.horizontalAlign === 'space-between' && isRowDirection && styles.justifyContentSpaceBetween,
    state.verticalAlign === 'space-between' && !isRowDirection && styles.justifyContentSpaceBetween,
    state.horizontalAlign === 'stretch' && !isRowDirection && styles.alignItemsStretch,
    state.verticalAlign === 'stretch' && isRowDirection && styles.alignItemsStretch,
    state.horizontalAlign === 'baseline' && !isRowDirection && styles.alignItemsBaseline,
    state.verticalAlign === 'baseline' && isRowDirection && styles.alignItemsBaseline,
    state.wrap === true && styles.flexWrap,
    state.wrap === false && styles.flexNoWrap,
    state.inline === true && styles.displayInlineFlex,
    state.inline === false && styles.displayFlex,
    state.className,
  );

  state.style = {
    ...state.style,
    [cssVars.gap]: state.gap === undefined ? 0 : state.gap,
    [cssVars.grow]: state.grow === undefined ? 0 : state.grow,
    [cssVars.shrink]: state.shrink === undefined ? 1 : state.shrink,
  };

  return state;
};
