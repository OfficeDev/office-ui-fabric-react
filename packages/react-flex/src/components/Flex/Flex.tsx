import * as React from 'react';
import { FlexProps, FlexSlots, FlexSlotProps } from './Flex.types';
import { compose, createClassResolver, mergeProps } from '@fluentui/react-compose';
import * as classes from './Flex.scss';
import { FlexItem } from '../FlexItem';
import { FlexContext } from './FlexContext';
import { useFlex } from './useFlex';

export const Flex = compose<'div', FlexProps, FlexProps, {}, {}>(
  (props, ref, options) => {
    const { children, disableShrink, tokens } = props;

    const { state } = options;
    const { slots, slotProps } = mergeProps<FlexProps, FlexProps, FlexSlots, FlexSlotProps>(state, options);

    const { flexChildren, generalChildren } = separateFlexChildren(children);

    return (
      <slots.root {...slotProps.root}>
        <FlexContext.Provider value={{ disableShrink: !!disableShrink, gap: tokens?.gap! && '0' }}>
          {flexChildren}
        </FlexContext.Provider>
        {generalChildren}
      </slots.root>
    );
  },
  {
    classes: createClassResolver(classes),
    displayName: 'Flex',
    handledProps: [
      'inline',
      'column',
      'wrap',
      'horizontalAlign',
      'verticalAlign',
      'reverse',
      'disableShrink',
      'fluid',
      'tokens',
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ] as any,
    state: useFlex,
  },
);

const separateFlexChildren = (children: React.ReactNode) => {
  const flexChildren: React.ReactNode[] = [];
  const generalChildren: React.ReactNode[] = [];

  React.Children.forEach(children, child => {
    if (isFlexItemElement(child)) {
      flexChildren.push(child);
    } else {
      generalChildren.push(child);
    }
  });

  return { flexChildren, generalChildren };
};

const isFlexItemElement = (item: React.ReactNode): item is typeof FlexItem => {
  return (
    !!item &&
    typeof item === 'object' &&
    !!(item as React.ReactElement).type &&
    ((item as React.ReactElement).type as React.ComponentType).displayName === FlexItem.displayName
  );
};

Flex.defaultProps = {
  as: 'div',
  inline: false,
  wrap: false,
  horizontalAlign: 'start',
  verticalAlign: 'stretch',
  disableShrink: false,
  fluid: false,
  tokens: {
    padding: '0',
    gap: '0',
  },
};
