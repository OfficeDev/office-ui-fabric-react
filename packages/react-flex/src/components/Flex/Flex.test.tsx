import * as React from 'react';
import { Flex } from './Flex';
import { render } from '@testing-library/react';
// import { isConformant } from '../../common/isConformant';
import { FlexDirectionProperty } from 'csstype';

describe('Flex', () => {
  // isConformant({
  //   Component: Flex,
  //   displayName: 'Flex',
  // });

  /**
   * Note: see more visual regression tests for Flex in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const utils = render(
      <Flex data-testid="container">
        <span>item</span>
      </Flex>,
    );

    const container = utils.getByTestId('container');
    expect(container).toHaveStyle({
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'normal',
      alignItems: 'normal',
      flexWrap: 'nowrap',
    });
  });

  it('renders default styles for immediate children', () => {
    const utils = render(
      <Flex>
        <span>affected</span>
        <span>
          <span>unaffected</span>
        </span>
      </Flex>,
    );

    const affected = utils.getByText('affected');
    expect(affected).toHaveStyle({
      margin: 0,
      order: 0,
      flexGrow: 0,
      flexShrink: 1,
      flexBasis: 'auto',
      alignSelf: 'auto',
    });

    const unaffected = utils.getByText('unaffected');
    expect(unaffected).not.toHaveStyle({
      margin: 0,
      order: 0,
      flexGrow: 0,
      flexShrink: 1,
      flexBasis: 'auto',
      alignSelf: 'auto',
    });
  });

  test.each([
    ['row', 'row'],
    ['row-reverse', 'row-reverse'],
    ['column', 'column'],
    ['column-reverse', 'column-reverse'],
  ])('changes direction to %s', (direction, expectation) => {
    const utils = render(
      <Flex direction={direction as FlexDirectionProperty} data-testid="container">
        <span>item</span>
      </Flex>,
    );

    const container = utils.getByTestId('container');

    expect(container).toHaveStyle({
      flexDirection: expectation,
    });
  });

  it('sets horizontal alginment', () => {
    const utils = render(
      <Flex direction="row" horizontalAlign="right" data-testid="container">
        <span>item</span>
      </Flex>,
    );

    const container = utils.getByTestId('container');

    expect(container).toHaveStyle({
      justifyContent: 'right',
    });
  });
});
