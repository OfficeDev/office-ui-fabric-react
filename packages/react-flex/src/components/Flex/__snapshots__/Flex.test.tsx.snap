// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex does not change node structure 1`] = `
<div>
  <div
    class=""
    style="display: flex; flex-direction: row; flex-wrap: nowrap; justify-content: flex-start; align-items: stretch;"
  >
    <div>
      Item 1
    </div>
    <div>
      Item 2
    </div>
  </div>
</div>
`;
